### ISSUE TITLE
Measuring Darkcount rate

### ISSUE THREAD

**Issue by dneise** [cta-sst-1m/digicampipe#176]
In PR #175 @calispac  wrote this:

Maybe to have better dark count rate measurement one could use Poisson count method (i.e. number of zero events). 
In this algorithm (`spe.py`) the zeros are not counted since we explicitly ask for finding pulses (above a certain thershold ~ 0.5 p.e.). What was told to me by @AndriiNagai is that we could just make a histogram of (`np.max(adc_samples, axis=-1) - np.min(adc_count, axis=-1)`).

This will not be so good for gain measurement since the 0. p.e peak is shifted to higher ADC count but would allow us to retrieve dark count rate. Also I doubt that the fit will work with the 0 p.e. peak. 

To create such a histogram as the code is, we need to implement a `find_pulse_x()` that fills `pulse_mask` with `np.arange(0, adc_samples.shape[-1]) == np.argmax(adc_samples, axis=-1)`
and have a function that fills the `baseline` Field with `np.min(adc_samples, axis=-1)`

----

And I thought this discussion is worth its own issue.

**Comment by dneise** [comment]
@AndriiNagai is completely right that using the so called "Poisson method" is a better way to estimate the dark count rate. 

I however have not yet heard of this particular implementation of this method. But I guess everything is worth a try.

**Comment by AndriiNagai** [comment]
Hi Dominik,

> On 8 Apr 2018, at 18:30, Dominik Neise <notifications@github.com> wrote:
> 
> @AndriiNagai <https://github.com/AndriiNagai> is completely right that using the so called "Poisson method" is a better way to estimate the dark count rate.
> 
> I however have not yet heard of this particular implementation of this method. But I guess everything is worth a try.
> 
I have used this method during my PhD. 
Please, find attached the comparison between DCR calculated from Poisson method and from measurements with Oscilloscope, where threshold was set to 0.5 p.e amplitude and time difference between two trigged events was used to calculate DCR.
You can see good agreement between both methods at T = 258.15K, while at higher T measurements with Oscilloscope saturated due to Oscilloscope dead time

Bets regards,
Andrii

**Comment by AndriiNagai** [comment]
This Poisson method is widely used during absolute PDE measurements. To correct the PDE to DCR, please see attached (Section 2.2).

Bets regards,
Andrii

**Comment by dneise** [comment]
@AndriiNagai  wrote:
> This Poisson method is widely used during absolute PDE measurements

Yes .. I absolutely agree. The Poisson method is widely used. But this particular implementation I have not heard of. By particular implementation, I mean using the difference between min and max a sampled timeline. I usually see the Poisson method being applied to a histogram which is the result of some kind charge extraction (usually integrating in a small time window around some kind of trigger, random trigger is also possible here).

That's all I'm saying. I did not want to criticize anything here. Neither the method, which is indeed well known, I think I saw it first 10 years ago used by Dieter Renker (the authot of this very nice overview article ..[Advances in solid state photon detectors](http://iopscience.iop.org/article/10.1088/1748-0221/4/04/P04004/pdf), but I am not sure where this method was used the first time ... I assume it is much older.

However, Dieter always integrated a bit, and did not use max-min.

The attachment unfortunately did not make it from the mail into this thread here. Maybe you could give us a link?

**Comment by AndriiNagai** [comment]
> On 9 Apr 2018, at 10:30, Dominik Neise <notifications@github.com> wrote:
> 
> @AndriiNagai <https://github.com/AndriiNagai> wrote:
> 
> This Poisson method is widely used during absolute PDE measurements
> 
> Yes .. I absolutely agree. The Poisson method is widely used. But this particular implementation I have not heard of. By particular implementation, I mean using the difference between min and max a sampled timeline. I usually see the Poisson method being applied to a histogram which is the result of some kind charge extraction (usually integrating in a small time window around some kind of trigger, random trigger is also possible here).
> 
> That's all I'm saying. I did not want to criticize anything here. Neither the method, which is indeed well known, I think I saw it first 10 years ago used by Dieter Renker (the authot of this very nice overview article ..Advances in solid state photon detectors <http://iopscience.iop.org/article/10.1088/1748-0221/4/04/P04004/pdf>, but I am not sure where this method was used the first time ... I assume it is much older.
> 
> However, Dieter always integrated a bit, and did not use max-min.
> 
For commercial SiPM devices, at a given T and Vbias, charge is proportional to amplitude. So, for me, between “integral" and "max-min" is only the difference that "max-min" (I would better write "max-baseline") is easier to calculate, because integral is more sensitive to baseline calculation.
> The attachment unfortunately did not make it from the mail into this thread here. Maybe you could give us a link?
> 
Sure: https://www.sciencedirect.com/science/article/pii/S0168900210008156 <https://www.sciencedirect.com/science/article/pii/S0168900210008156>

May be, we can discuss this during the meeting?

Best regards,
Andrii

**Comment by calispac** [comment]
Awesome ! Could you tell me what is random charge method ? 
Also If we can merge what you have done in #182 that would be nice.

**Comment by dneise** [comment]
> Could you tell me what is random charge method ?

Sure sorry.

Imagine this lab setup:  
 * you have pulsed light source with a very low expectation value for the number of photons, e.g. 0.1 photon per shot.
 * you have your PMT/SiPM, a pre-amp and a normal oscilloscope. 
 * you trigger the scope with the trigger from the light source. 
 * at the trigger time you integrate the signal on the scope for ~5ns or so (whatever the typical pulse width is)

Then you get a similar spectrum as we get with our SPE-analysis, but since your light source is so weak, you will very often get a trigger put no pulse at all, resulting in a very high 0-p.e. peak in the spectrum.

From this 0-p.e. peak in the spectrum one can nicely estimate the expectation value of the light source.

-----

Now with dark counts, we do not have a trigger, but we can simply think of dark counts as a weak light source. In any given integration window (of e.g. 10ns width) the probability of having a dark count is very low (at 3MHz it is 3%). So we can simply create a random trigger, and make the same measurement as the one we did in the lab with a triggered light source. 

---- 

This is basically the "Poisson method" as far as I understood it.

Now the analysis goes like this:

* count the number of entries in the 0-p.e. peak:  N0
* count the total number of entries. N
* frequentist probability for "no pulse in time window" P0 = N0 / N
* Poisson probability for "exactly zero events in window" = exp(- mu) 
* where mu is the expectation value: mu = time_window * average_rate

So one can calculate the rate as: 

average_rate = - ln(N0/N) / time_window

**Comment by calispac** [comment]
Ok so this means that we are underestimating the N0 if the dark count rate is overestimated ? 

This seems normal since the 0 p.e. peak and the 1 p.e. peak are not well separated.

**Comment by dneise** [comment]
Yes, underestimating N0 is one possibility or overestimating N.

How was this spectrum made?

**Comment by AndriiNagai** [comment]
Nice work Dominik!

From my experience, the tricky point is to separate 0p.e. from 1p.e. peak. 
Typically, the histogram contains entries between 0 p.e. and 1 p.e. how did the algorithm decide if entry with 0.5 p.e amplitude go to N0 or not? Could you show the example of max - min distribution? Is it possible to use something like: “max - baseline”, where baseline is an average over given time window before t_max - SiPM_risetime?

Another, very stupid question: Are you sure that you generated exactly 3MHz of DCR? Not 3.1 or 2.9? Could your waveform contain only part of SiPM pulses (if pulse was generated in the beginning or in the end of waveform)?

Best regards,
Andrii

**Comment by dneise** [comment]
Unortunately I have no time to further go into details with this study at the moment. But all the details can be found here.

https://github.com/dneise/study_cta_sst1m_darkcount_rate

I might find time to add a readme.

**Comment by calispac** [comment]
> average_rate = - ln(N0/N) / time_window

@dneise  I just realized that the time_window is not 50 ns (e.g. time of readout window) but rather an effective window on which pulses are search for. In this function : 
https://github.com/cta-sst-1m/digicampipe/blob/4ab7ae2c22387a0d9c0ec5214d03d262e1703631/digicampipe/scripts/spe.py#L215-L237

Pulses are searched over : T = n_samples * 4 ns = (50 - (2*6 + 1)) *4 = 148 ns. 
Is that correct ?

**Comment by calispac** [comment]
Ah sorry... This has to do with the filter that is applied...

**Comment by calispac** [comment]
Maybe one could use this has a smoothing function : 
https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.gaussian_filter1d.html#scipy.ndimage.gaussian_filter1d

**Comment by dneise** [comment]
from: dark_3MHz_toymc.h5.h5

![image](https://user-images.githubusercontent.com/8200858/38621963-5caa76d6-3da2-11e8-9c2a-852479ecaae9.png)

**Comment by calispac** [comment]
See this commit https://github.com/cta-sst-1m/digicampipe/pull/182/commits/5279463215014e9ff97766f5e69d4a38132018e4

Where I rewrote the pulse finder with gaussian filter.

**Comment by dneise** [comment]
Cyril .. I am not sure I can follow you. What is in your opinion the problem?

**Comment by calispac** [comment]
> Cyril .. I am not sure I can follow you. What is in your opinion the problem?

The problem is that right now in find_pulse_3() pulses are found only if they are within `[:, 6:-6]` therefore we are throwing away some data. But I shouldn't have posted that here since the max-min method is not affected by this

**Comment by dneise** [comment]
I agree, neither the max_min method, nor the random_trigger method is affected by this.
